The goal of this notebook is to test the modularity data analysis pipeline on a test case. 

In [1]:
from neuprint import Client
# remove my token before making notebook public
c = Client('neuprint.janelia.org', dataset='hemibrain:v1.2.1', token='eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImdnMjExNEBjb2x1bWJpYS5lZHUiLCJsZXZlbCI6Im5vYXV0aCIsImltYWdlLXVybCI6Imh0dHBzOi8vbGgzLmdvb2dsZXVzZXJjb250ZW50LmNvbS9hLS9BT2gxNEdpb1lJLUVPLWdidGxPRTh6SmQ0eF9ZQ1Y4ZHF0YVFjWGlHeG5CMz1zOTYtYz9zej01MD9zej01MCIsImV4cCI6MTgxMDUyOTYzNH0.jv9eR0SH5RhfBdXrtp4r-dDFOhcsT8GBbE4v69ysCKs') 
c.fetch_version()

# import important stuff here
import numpy as np
import pandas as pd
import matplotlib

In [2]:
# testing on a neuron with few connections
test_neuron_Id = 676124666# 1815929980 #450971893 #423101189 #

In [3]:
from neuprint import fetch_neurons
test_neuron_df, test_syns = fetch_neurons(test_neuron_Id)
test_neuron_df

bodyId instance  type  pre  post  downstream  upstream  mito    size  \
0  676124666     None  None    1     1           9         1     0  178283   

  status cropped statusLabel cellBodyFiber somaRadius somaLocation  \
0   None    None        None          None       None         None   

                                             roiInfo notes         inputRois  \
0  {'SNP(R)': {'pre': 1, 'post': 1, 'downstream':...  None  [SMP(R), SNP(R)]   

         outputRois  
0  [SMP(R), SNP(R)]

In [4]:
test_syns

bodyId         roi  pre  post  downstream  upstream  mito
0  676124666  NotPrimary    0     0           0         0     0
1  676124666      SMP(R)    1     1           9         1     0
2  676124666      SNP(R)    1     1           9         1     0

In [5]:
from neuprint import fetch_simple_connections

test_inputs = fetch_simple_connections(None, test_neuron_Id)
test_inputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0   644421171    676124666       1     None      None         None   

  instance_post                                       conn_roiInfo  
0          None  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...

All of these downstream neurons are contacted from the same pre site. Interestingly, it seems that weight is not necessarily a count of the post sites as I assumed since the weight would've added up to the downstream total. This is true regardless of whether fetch_simple_connections or fetch_adjacencies is used. This is important to note for other stuff. Even when include_nonprimary=True for fetch_adjacencies, there are only 8 entries and it's just because synapses are being double counted in SMP and SNP rois.

In [6]:
from neuprint import fetch_simple_connections

test_outputs = fetch_simple_connections(test_neuron_Id,None)
test_outputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0   676124666    423101189       1     None     oviIN         None   
1   676124666    644761952       1     None   LHPD2c5         None   
2   676124666   5813009620       1     None    SMP109         None   
3   676124666   5813056054       1     None    SMP050         None   

  instance_post                                       conn_roiInfo  
0       oviIN_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
1     LHPD2c5_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
2      SMP109_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
3      SMP050_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...

In [7]:
from neuprint import fetch_adjacencies

[testn_, test_outfetch] = fetch_adjacencies(test_neuron_Id,None)
test_outfetch

bodyId_pre  bodyId_post     roi  weight
0   676124666    423101189  SMP(R)       1
1   676124666    644761952  SMP(R)       1
2   676124666   5813009620  SMP(R)       1
3   676124666   5813056054  SMP(R)       1

In [8]:
from neuprint import fetch_adjacencies

[testn_, test_outfetch] = fetch_adjacencies(test_neuron_Id,None, include_nonprimary=True)
test_outfetch

bodyId_pre  bodyId_post     roi  weight
0   676124666    423101189  SNP(R)       1
1   676124666    423101189  SMP(R)       1
2   676124666    644761952  SNP(R)       1
3   676124666    644761952  SMP(R)       1
4   676124666   5813009620  SNP(R)       1
5   676124666   5813009620  SMP(R)       1
6   676124666   5813056054  SNP(R)       1
7   676124666   5813056054  SMP(R)       1

Just to harp on this point, this particular neuron (id: 676124666) has 1 pre site and 1 post site on its body. Each of its 4 downstream partners is connecting to the same pre site. You can see this when you fetch_synapse_connections for this neuron since all the pre sites have the exact same coordinates. It is entirely unclear what the "downstream" count of 9 is referring to. I will have to revisit anywhere we have used the downstream and upstream information and double check those results.

In [9]:
from neuprint import fetch_synapses

fetch_synapses(test_neuron_Id)

bodyId  type     roi      x      y     z  confidence
0  676124666   pre  SMP(R)  18027  31040  9294    0.902000
1  676124666  post  SMP(R)  18058  31070  9291    0.677568

In [10]:
from neuprint import fetch_synapse_connections
fetch_synapse_connections(test_neuron_Id,None)

  0%|          | 0/4 [00:00<?, ?it/s]

bodyId_pre  bodyId_post roi_pre roi_post  x_pre  y_pre  z_pre  x_post  \
0   676124666   5813056054  SMP(R)   SMP(R)  18027  31040   9294   17990   
1   676124666    423101189  SMP(R)   SMP(R)  18027  31040   9294   18025   
2   676124666    644761952  SMP(R)   SMP(R)  18027  31040   9294   18039   
3   676124666   5813009620  SMP(R)   SMP(R)  18027  31040   9294   18030   

   y_post  z_post  confidence_pre  confidence_post  
0   31030    9254           0.902            0.936  
1   31034    9274           0.902            0.994  
2   31043    9274           0.902            0.995  
3   31020    9323           0.902            0.370

To move on from this issue, I next test out the different options for the get_connectome function. Below is the default.
# Testing get_connectome

In [11]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id)
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644421171       9
1    423101189    644761952      59
2    423101189   5813009620      15
3    423101189   5813056054      78
4    644421171    423101189      51
5    644421171    644761952       3
6    644421171    676124666       1
7    644421171   5813009620       1
8    644421171   5813056054       1
9    644761952    423101189      20
10   644761952   5813009620      18
11   644761952   5813056054       5
12   676124666    423101189       1
13   676124666    644761952       1
14   676124666   5813009620       1
15   676124666   5813056054       1
16  5813009620    423101189      92
17  5813009620   5813056054      12
18  5813056054    423101189       2
19  5813056054   5813009620       1

This was done in the function by fetching adjacencies for inputs and outputs of the main neurons.

If we exclude the main neuron, you can count that 5 rows from above should be excluded and this is indeed the result below.

In [12]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, exclude_main_neurons=True)
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644421171       9
1    423101189    644761952      59
2    423101189   5813009620      15
3    423101189   5813056054      78
4    644421171    423101189      51
5    644421171    644761952       3
6    644421171   5813009620       1
7    644421171   5813056054       1
8    644761952    423101189      20
9    644761952   5813009620      18
10   644761952   5813056054       5
11  5813009620    423101189      92
12  5813009620   5813056054      12
13  5813056054    423101189       2
14  5813056054   5813009620       1

If we change the scope of the connectome to be inputs only, we get the only input connection which is with the main neuron itself.

In [14]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, connectome_scope='input')
test_connectome

bodyId_pre  bodyId_post  weight
0   644421171    676124666       1

Conversely, changing the scope to output only excludes the 6 rows where that only input neuron (id: 644421171) appeared.

In [15]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, connectome_scope='output')
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644761952      59
1    423101189   5813009620      15
2    423101189   5813056054      78
3    644761952    423101189      20
4    644761952   5813009620      18
5    644761952   5813056054       5
6    676124666    423101189       1
7    676124666    644761952       1
8    676124666   5813009620       1
9    676124666   5813056054       1
10  5813009620    423101189      92
11  5813009620   5813056054      12
12  5813056054    423101189       2
13  5813056054   5813009620       1

To test the thresholding by weight and the connectome by type options, we'll need a different test case that has multiple instances of those things. I'll only do this on the input scope to limit the number of cell types to count up.

In [34]:
# testing on a neuron with few connections
test_neuron_Id = 547612423 #5813077939 #7112615328#547612423#

In [3]:
from neuprint import fetch_neurons
test_neuron_df, test_syns = fetch_neurons(test_neuron_Id)
test_neuron_df

bodyId   instance  type  pre  post  downstream  upstream  mito  \
0  547612423  (ADM04)_L  None   78   107         692       107    50   

        size  status  cropped     statusLabel cellBodyFiber somaRadius  \
0  191705400  Traced    False  Roughly traced          None       None   

  somaLocation                                            roiInfo notes  \
0         None  {'SNP(L)': {'pre': 55, 'post': 80, 'downstream...  None   

                          inputRois                        outputRois  
0  [SMP(L), SMP(R), SNP(L), SNP(R)]  [SMP(L), SMP(R), SNP(L), SNP(R)]

In [18]:
from neuprint import fetch_simple_connections

test_inputs = fetch_simple_connections(None, test_neuron_Id)
test_inputs

bodyId_pre  bodyId_post  weight type_pre type_post        instance_pre  \
0    298254517    547612423       7   SMP082      None     SMP082(ADM04)_L   
1    516943720    547612423       5     None      None  (aDT6f)(MBDLaxon2)   
2   1141035086    547612423       4     None      None  (aDT6f)(MBDLaxon2)   
3   5813055769    547612423       4   SLP212      None     SLP212(ADL27)_L   
4    328265389    547612423       3   SMP082      None            SMP082_R   
5    329310230    547612423       3   SMP082      None     SMP082(ADM04)_L   
6    359749303    547612423       3     None      None                None   
7    455500416    547612423       3    PAL02      None      PAL02(ADL27)_L   
8    550029475    547612423       3     None      None         (MBDLaxon2)   
9   5812980467    547612423       3     None      None                None   
10  5813013636    547612423       3     None      None  (aDT6a)(MBDLaxon2)   
11   297925526    547612423       2     None      None                None   
12   328373131    547612423       2   SMP286      None     SMP286(PDM05)_L   
13   360427529    547612423       2     None      None                None   
14   361445621    547612423       2   SMP108      None     SMP108(ADM05)_L   
15   391802788    547612423       2     None      None                None   
16   486255488    547612423       2     None      None  (aDT6b)(MBDLaxon2)   
17   486544396    547612423       2    PAL02      None             PAL02_R   
18   579014562    547612423       2   SMP103      None            SMP103_R   
19   611728142    547612423       2     None      None  (aDT6a)(MBDLaxon2)   
20   674501939    547612423       2     None      None         (MBDLaxon2)   
21  5813026589    547612423       2   SMP082      None            SMP082_R   
22  5813133687    547612423       2     None      None                None   
23  5901203505    547612423       2   SLP212      None            SLP212_R   
24   267214250    547612423       1     pC1b      None              pC1b_R   
25   328632707    547612423       1   SMP083      None            SMP083_R   
26   390706983    547612423       1   SMP231      None            SMP231_R   
27   392325925    547612423       1   SMP103      None            SMP103_R   
28   393503882    547612423       1     None      None                None   
29   422487932    547612423       1     None      None                None   
30   424889074    547612423       1     None      None                None   
31   448260940    547612423       1      LNd      None               LNd_R   
32   453535678    547612423       1     None      None                None   
33   453877678    547612423       1     None      None                None   
34   478937243    547612423       1   SMP228      None            SMP228_R   
35   487610557    547612423       1   SMP094      None     SMP094(ADM04)_L   
36   513848395    547612423       1     None      None         (MBDLaxon2)   
37   515524209    547612423       1     None      None         (MBDLaxon1)   
38   519603027    547612423       1     None      None                None   
39   545621597    547612423       1     None      None           (PDM15)_L   
40   550322829    547612423       1   SMP602      None    SMP602(SCB061)_L   
41   551666144    547612423       1   SMP106      None            SMP106_R   
42   580010810    547612423       1     None      None  (aDT6f)(MBDLaxon2)   
43   581013183    547612423       1   aSP-f2      None            aSP-f2_R   
44   581013253    547612423       1     None      None  (aDT6a)(MBDLaxon2)   
45   581409749    547612423       1   SMP093      None     SMP093(ADM04)_L   
46   582372162    547612423       1     None      None                None   
47   610049447    547612423       1     None      None                None   
48   643039876    547612423       1     None      None           (PVM12)_L   
49   643462210    547612423       1     None      None                None   
50   644861161    54761242

In [4]:
from neuprint import fetch_adjacencies

test_pre, _ = fetch_adjacencies(None, test_neuron_Id)
test_pre

bodyId    type         instance
0    547612423    None        (ADM04)_L
1    267214250    pC1b           pC1b_R
2    297925526    None             None
3    298254517  SMP082  SMP082(ADM04)_L
4    328265389  SMP082         SMP082_R
..         ...     ...              ...
56  5813046657    None        (PDM12)_L
57  5813055769  SLP212  SLP212(ADL27)_L
58  5813133687    None             None
59  5901203505  SLP212         SLP212_R
60  5901212536    None             None

[61 rows x 3 columns]

In [61]:
test_pre = test_pre[~test_pre.bodyId.isin([test_neuron_Id])]
test_pre

bodyId    type            instance
1    267214250    pC1b              pC1b_R
2    297925526    None                None
3    298254517  SMP082     SMP082(ADM04)_L
4    328265389  SMP082            SMP082_R
5    328373131  SMP286     SMP286(PDM05)_L
6    328632707  SMP083            SMP083_R
7    329310230  SMP082     SMP082(ADM04)_L
8    359749303    None                None
9    360427529    None                None
10   361445621  SMP108     SMP108(ADM05)_L
11   390706983  SMP231            SMP231_R
12   391802788    None                None
13   392325925  SMP103            SMP103_R
14   393503882    None                None
15   422487932    None                None
16   424889074    None                None
17   448260940     LNd               LNd_R
18   453535678    None                None
19   453877678    None                None
20   455500416   PAL02      PAL02(ADL27)_L
21   478937243  SMP228            SMP228_R
22   486255488    None  (aDT6b)(MBDLaxon2)
23   486544396   PAL02             PAL02_R
24   487610557  SMP094     SMP094(ADM04)_L
25   513848395    None         (MBDLaxon2)
26   515524209    None         (MBDLaxon1)
27   516943720    None  (aDT6f)(MBDLaxon2)
28   519603027    None                None
29   545621597    None           (PDM15)_L
30   550029475    None         (MBDLaxon2)
31   550322829  SMP602    SMP602(SCB061)_L
32   551666144  SMP106            SMP106_R
33   579014562  SMP103            SMP103_R
34   580010810    None  (aDT6f)(MBDLaxon2)
35   581013183  aSP-f2            aSP-f2_R
36   581013253    None  (aDT6a)(MBDLaxon2)
37   581409749  SMP093     SMP093(ADM04)_L
38   582372162    None                None
39   610049447    None                None
40   611728142    None  (aDT6a)(MBDLaxon2)
41   643039876    None           (PVM12)_L
42   643462210    None                None
43   644861161    None                None
44   674501939    None         (MBDLaxon2)
45   704487840  SMP602    SMP602(SCB061)_R
46  1141035086    None  (aDT6f)(MBDLaxon2)
47  5812980467    None                None
48  5812980529  SLP212            SLP212_R
49  5813009950    None         (MBDLaxon2)
50  5813011454    None           (PDM12)_L
51  5813012274    None         (MBDLaxon2)
52  5813013636    None  (aDT6a)(MBDLaxon2)
53  5813026589  SMP082            SMP082_R
54  5813026705  SMP589            SMP589_R
55  5813035328  SMP525            SMP525_R
56  5813046657    None           (PDM12)_L
57  5813055769  SLP212     SLP212(ADL27)_L
58  5813133687    None                None
59  5901203505  SLP212            SLP212_R
60  5901212536    None                None

In [11]:
from neuprint import fetch_adjacencies

# get connections among neurons using the bodyIds from pre
partners_, t_connectome = fetch_adjacencies(test_pre['bodyId'], test_pre['bodyId'])
t_connectome = t_connectome.groupby(['bodyId_pre', 'bodyId_post'], as_index=False)['weight'].sum()
t_connectome

bodyId_pre  bodyId_post  weight
0     267214250    297925526       2
1     267214250    298254517       2
2     267214250    328265389       2
3     267214250    328373131       2
4     267214250    328632707       1
..          ...          ...     ...
810  5901203505   5813009950       3
811  5901203505   5813012274      14
812  5901203505   5813013636       9
813  5901203505   5813055769       5
814  5901212536    547612423       1

[815 rows x 3 columns]

In [12]:
weight_thresh = 25
t_connectome = t_connectome[t_connectome['weight'] >= weight_thresh]
t_connectome

bodyId_pre  bodyId_post  weight
21    297925526    486544396      29
34    298254517    328632707      36
56    328265389    328632707      44
101   328632707    328373131      38
112   329310230    328632707      26
246   455500416    486544396      33
296   486544396    455500416      30
315   486544396   5813055769      26
382   519603027   5812980467      43
383   519603027   5812980529      40
384   519603027   5813055769      32
417   550029475    486544396      26
514   582372162    486544396      41
536   611728142    455500416      34
537   611728142    486544396      40
705  5813026589    328632707      51
733  5813026705    455500416      37
734  5813026705    486544396      39
780  5813055769    455500416      28
781  5813055769    486544396      37
792  5813055769   5812980529      25
801  5901203505    455500416      90
802  5901203505    486544396      88

In [13]:
t_connectome = t_connectome.merge(partners_[['bodyId','type']], left_on='bodyId_pre', right_on='bodyId').rename(columns={'type':'type_pre'})

# merge type_post information from partners_ into connectome and rename type column to type_post
t_connectome = t_connectome.merge(partners_[['bodyId','type']], left_on='bodyId_post', right_on='bodyId').rename(columns={'type':'type_post'})
t_connectome

bodyId_pre  bodyId_post  weight    bodyId_x type_pre    bodyId_y type_post
0    297925526    486544396      29   297925526     None   486544396     PAL02
1    455500416    486544396      33   455500416    PAL02   486544396     PAL02
2    550029475    486544396      26   550029475     None   486544396     PAL02
3    582372162    486544396      41   582372162     None   486544396     PAL02
4    611728142    486544396      40   611728142     None   486544396     PAL02
5   5813026705    486544396      39  5813026705   SMP589   486544396     PAL02
6   5813055769    486544396      37  5813055769   SLP212   486544396     PAL02
7   5901203505    486544396      88  5901203505   SLP212   486544396     PAL02
8    298254517    328632707      36   298254517   SMP082   328632707    SMP083
9    328265389    328632707      44   328265389   SMP082   328632707    SMP083
10   329310230    328632707      26   329310230   SMP082   328632707    SMP083
11  5813026589    328632707      51  5813026589   SMP082   328632707    SMP083
12   328632707    328373131      38   328632707   SMP083   328373131    SMP286
13   486544396    455500416      30   486544396    PAL02   455500416     PAL02
14   611728142    455500416      34   611728142     None   455500416     PAL02
15  5813026705    455500416      37  5813026705   SMP589   455500416     PAL02
16  5813055769    455500416      28  5813055769   SLP212   455500416     PAL02
17  5901203505    455500416      90  5901203505   SLP212   455500416     PAL02
18   486544396   5813055769      26   486544396    PAL02  5813055769    SLP212
19   519603027   5813055769      32   519603027     None  5813055769    SLP212
20   519603027   5812980467      43   519603027     None  5812980467      None
21   519603027   5812980529      40   519603027     None  5812980529    SLP212
22  5813055769   5812980529      25  5813055769   SLP212  5812980529    SLP212

I'm so glad I checked this because it turns out that groupby drops any 'None'. I need to replace None with a string.

In [14]:
# replace None values in type_pre and type_post with 'unspecified'
t_connectome['type_pre'] = t_connectome['type_pre'].fillna('unspecified')
t_connectome['type_post'] = t_connectome['type_post'].fillna('unspecified')
t_connectome

bodyId_pre  bodyId_post  weight    bodyId_x     type_pre    bodyId_y  \
0    297925526    486544396      29   297925526  unspecified   486544396   
1    455500416    486544396      33   455500416        PAL02   486544396   
2    550029475    486544396      26   550029475  unspecified   486544396   
3    582372162    486544396      41   582372162  unspecified   486544396   
4    611728142    486544396      40   611728142  unspecified   486544396   
5   5813026705    486544396      39  5813026705       SMP589   486544396   
6   5813055769    486544396      37  5813055769       SLP212   486544396   
7   5901203505    486544396      88  5901203505       SLP212   486544396   
8    298254517    328632707      36   298254517       SMP082   328632707   
9    328265389    328632707      44   328265389       SMP082   328632707   
10   329310230    328632707      26   329310230       SMP082   328632707   
11  5813026589    328632707      51  5813026589       SMP082   328632707   
12   328632707    328373131      38   328632707       SMP083   328373131   
13   486544396    455500416      30   486544396        PAL02   455500416   
14   611728142    455500416      34   611728142  unspecified   455500416   
15  5813026705    455500416      37  5813026705       SMP589   455500416   
16  5813055769    455500416      28  5813055769       SLP212   455500416   
17  5901203505    455500416      90  5901203505       SLP212   455500416   
18   486544396   5813055769      26   486544396        PAL02  5813055769   
19   519603027   5813055769      32   519603027  unspecified  5813055769   
20   519603027   5812980467      43   519603027  unspecified  5812980467   
21   519603027   5812980529      40   519603027  unspecified  5812980529   
22  5813055769   5812980529      25  5813055769       SLP212  5812980529   

      type_post  
0         PAL02  
1         PAL02  
2         PAL02  
3         PAL02  
4         PAL02  
5         PAL02  
6         PAL02  
7         PAL02  
8        SMP083  
9        SMP083  
10       SMP083  
11       SMP083  
12       SMP286  
13        PAL02  
14        PAL02  
15        PAL02  
16        PAL02  
17        PAL02  
18       SLP212  
19       SLP212  
20  unspecified  
21       SLP212  
22       SLP212

In [15]:
# group by type_pre and type_post and sum the weights
t_connectome = t_connectome[['type_pre','type_post','weight']].groupby(['type_pre','type_post'], as_index=False).sum()
t_connectome

type_pre    type_post  weight
0        PAL02        PAL02      63
1        PAL02       SLP212      26
2       SLP212        PAL02     243
3       SLP212       SLP212      25
4       SMP082       SMP083     157
5       SMP083       SMP286      38
6       SMP589        PAL02      76
7  unspecified        PAL02     170
8  unspecified       SLP212      72
9  unspecified  unspecified      43

In [35]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, exclude_main_neurons=True, connectome_scope='input', weight_threshold=weight_thresh, connectome_by_type=True)
test_connectome

type_pre    type_post  weight
0        PAL02        PAL02      63
1        PAL02       SLP212      26
2       SLP212        PAL02     243
3       SLP212       SLP212      25
4       SMP082       SMP083     157
5       SMP083       SMP286      38
6       SMP589        PAL02      76
7  unspecified        PAL02     170
8  unspecified       SLP212      72
9  unspecified  unspecified      43

Now to check that connectome_to_undirected works as expected. There is only 1 reciprocal connection in the above example, so the function will return 1 less row.

In [20]:
undirected_edges = {}  # Dictionary to store the undirected edges and their weights

# Determine the column names for the pre and post neurons
connectome_columns = test_connectome.columns
pre = [col for col in connectome_columns if 'pre' in col][0]
post = [col for col in connectome_columns if 'post' in col][0]

for index, row in test_connectome.iterrows():
    source = row[pre]
    target = row[post]
    weight = row['weight']

    # Check if the edge already exists in the reverse
    if (target, source) in undirected_edges:
        # Update the weight of the existing edge
        undirected_edges[(target, source)] += weight
    else:
        # Add a new edge to dict
        undirected_edges[(source, target)] = weight

undirected_edges

{('PAL02', 'PAL02'): 63,
 ('PAL02', 'SLP212'): 269,
 ('SLP212', 'SLP212'): 25,
 ('SMP082', 'SMP083'): 157,
 ('SMP083', 'SMP286'): 38,
 ('SMP589', 'PAL02'): 76,
 ('unspecified', 'PAL02'): 170,
 ('unspecified', 'SLP212'): 72,
 ('unspecified', 'unspecified'): 43}

In [21]:
# Create a DataFrame from the undirected edges dictionary
undirected_edgelist = pd.DataFrame(list(undirected_edges.keys()), columns=['source', 'target'])
undirected_edgelist['weight'] = list(undirected_edges.values())
undirected_edgelist

source       target  weight
0        PAL02        PAL02      63
1        PAL02       SLP212     269
2       SLP212       SLP212      25
3       SMP082       SMP083     157
4       SMP083       SMP286      38
5       SMP589        PAL02      76
6  unspecified        PAL02     170
7  unspecified       SLP212      72
8  unspecified  unspecified      43

In [36]:
from get_connectome import connectome_to_undirected

test_connectome_undirected = connectome_to_undirected(test_connectome)
test_connectome_undirected

source       target  weight
0        PAL02        PAL02      63
1        PAL02       SLP212     269
2       SLP212       SLP212      25
3       SMP082       SMP083     157
4       SMP083       SMP286      38
5       SMP589        PAL02      76
6  unspecified        PAL02     170
7  unspecified       SLP212      72
8  unspecified  unspecified      43

When creating a connectome by cell type, the nodes need to be numbers and not strings. I thought that format_edgelist and the C code would automatically deal with this, but those didn't work on the connectomes by cell types. The key file never gets made. Rhessa has some code to assign 8-digit numbers to cell types. I will test that out below.

# Official start of analysis pipeline
Now that I got to see exactly what this test neuron is connected to, the official analysis pipeline starts below. 

I use the get_connectome function to get a dataframe with the neuron's connectome. I want connectome_to_undirected to be a function that can be called separately and that returns the directed connectome dataframe. It is useful for these to be separate so that we can do other analyses with them besides modularity. 

In [38]:
# testing on a neuron with few connections
test_neuron_Id = 676124666# 1815929980 #450971893 #423101189 #

In [39]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id)
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644421171       9
1    423101189    644761952      59
2    423101189   5813009620      15
3    423101189   5813056054      78
4    644421171    423101189      51
5    644421171    644761952       3
6    644421171    676124666       1
7    644421171   5813009620       1
8    644421171   5813056054       1
9    644761952    423101189      20
10   644761952   5813009620      18
11   644761952   5813056054       5
12   676124666    423101189       1
13   676124666    644761952       1
14   676124666   5813009620       1
15   676124666   5813056054       1
16  5813009620    423101189      92
17  5813009620   5813056054      12
18  5813056054    423101189       2
19  5813056054   5813009620       1

In [40]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

test_connectome_undirected = connectome_to_undirected(test_connectome)
test_connectome_undirected

source      target  weight
0    423101189   644421171      60
1    423101189   644761952      79
2    423101189  5813009620     107
3    423101189  5813056054      80
4    644421171   644761952       3
5    644421171   676124666       1
6    644421171  5813009620       1
7    644421171  5813056054       1
8    644761952  5813009620      18
9    644761952  5813056054       5
10   676124666   423101189       1
11   676124666   644761952       1
12   676124666  5813009620       1
13   676124666  5813056054       1
14  5813009620  5813056054      13

The undirected connectome dataframe is ready to be exported to a csv file. From there it will be formatted using format_edgelist.py from the terminal window so that the nodes are no longer bodyIds but sequential numbers instead. Then the modularity should be able to run. It turns out that Alex's newest version of format_edgelist.py also makes the connectome undirected, so running the connectome_to_undirected function before format_edgelist.py is not necessary but also seems to be harmless. I verified that the results are the same regardless of whether connectome_to_undirected is run or not.

When exporting, either remove the header row or make sure to use the header flag when calling format_edgelist.py. Appropriate results were obtained with
`python format_edgelist.py test_connectome_undirected.txt --sep comma` in the terminal window using the exported file created below. test_connectome_undirected.txt gets overwritten and replaced by the sequentially numbered nodes. A copy of the original is created that is called original_test_connectome_undirected.txt, as well as a few other files (clean_, degree_, info_, and key_ files).

The exported file can be .txt or .csv. The files that are created from format_edgelist.py will inherit the same file type. I chose to export as .txt.

In [41]:
# export the undirected connectome to a csv or txt file
test_connectome_undirected.to_csv('test_connectome_undirected.txt', index = False, header=False)
#test_connectome_undirected.to_csv('for_running_modularity/test/test_connectome_undirected.txt', index = False, header=False)
#test_connectome_undirected.to_csv('test_connectome_undirected.txt', index = False, header=False, sep=' ')

After running format_edgelist.py in the terminal window, I then run: `sh work.sh test_connectome_undirected.txt` but I think this actually generates the same files that Alex generates with his new format_edgelist.py. The only problem with going directly to work.sh after saving test_connectome_undirected.txt is that work.sh doesn't renumber the nodes which we need. So effectively, format_edgelist.py rolls that step into the work.sh step all in one go. Alternatively, we could just run the functions separately from format_edgelist.py in this notebook. To do that, I had to change the separator to ' ' when exporting 'test_connectome_undirected.txt' in order to avoid errors. Below I made an attempt but the resulting new_test_connectome_undirected.txt file had renumbered the nodes correctly but the numbers were not sequential which was odd. I think it would be better to run `format_edgelist(input)` in a notebook where the input is a struct that contains prefix, suffix, and file path as far as I can tell. This would run all the things the way that running format_edgelist.py in the terminal does. I can play around with that if needed but I'll move on since we have a working solution which is to simply run format_edgelist.py in the terminal.

In [ ]:
# alternative to running format_edgelist.py from command line in terminal window
from format_edgelist import read_graph
from format_edgelist import write_edges
from pathlib import Path

myfile = Path('test_connectome_undirected.txt')
nodes, degrees, edges = read_graph(myfile)

myoutfile = Path('new_test_connectome_undirected.txt')
write_edges(myoutfile, edges, nodes)

If I run `python format_edgelist.py test_connectome_undirected.txt --sep comma` in the terminal window and follow that with `clang main.c help.c rg.c -Xpreprocessor -fopenmp -lomp -lm` and `./a.out 2 5 2 12345 0 test_connectome_undirected.txt`, the results_ file is different than when I include the `sh work.sh test_connectome_undirected.txt` step. In the test that I did, the results ended up with more communities and slightly different numbers of neurons in them. It is unclear to me which is more correct. If I run the `sh work.sh test_connectome_undirected.txt` step twice, I don't get different results from when I run it once, so that makes me think that the reason results are different when I don't run it is because it does something that `python format_edgelist.py test_connectome_undirected.txt --sep comma` doesn't do. So I will continue including the `sh work.sh test_connectome_undirected.txt` step even though it is partially redundant with some of what format_edgelist.py does.

The pipeline so far is:

First run `python format_edgelist.py test_connectome_undirected.txt --sep comma` in the terminal window. test_connectome_undirected.txt is overwritten with an undirected edge list that contains sequentially numbered nodes. A new file appears: original_, along with clean_, key_, info_, degree_ if all is going well. Some of these files might not appear if the nodes are labeled with strings instead of numbers and we need to fix that.

Then run `sh work.sh test_connectome_undirected.txt`. After running, the required files appear:, clean_, key_, info_, degree_ (if format_edgelist.py wasn't used to create them) and we next compile the C codes. I had to modify Pramesh's command to use clang given the weird compatibility issues with new macs and gcc and CMake.

`clang main.c help.c rg.c -Xpreprocessor -fopenmp -lomp -lm` to generate the a.out file. On my computer I get a bunch of warnings but none seems important.


`./a.out 2 5 2 12345 0 test_connectome_undirected.txt` is supposed to generate results_test_connectome_undirected.txt and partition_test_connectome_undirected.txt, although in these test cases that partition_ file didn't appear. In fact, it seems like the C program doesn't finish running even when I use a tiny connectome.

When I tested this out on oviIN_R's full connectome with chi=0 I thought I ended up with 181 clusters but it turns out, those numbers are iterations of the algorithm. It seems like the algorithm doesn't finish running on my computer for some reason because it is supposed to end with a final Q value. 

I might be using improper parameters for the CPUs, etc but Rhessa uses the same default CPU parameters and even the same random seed. I suspect that something weird is going on with a library somewhere since the same thing was happening on the lab computer and even compiling the codes was giving me problems. 

Sidenote: When trying to do this on the lab computer, I had to compile the code using a bunch of additional flags so that the libomp library could be accessed since it's not in the place where clang expects: `clang -I/Users/ggutierr/miniconda3/pkgs/llvm-openmp-14.0.6-hc6e5704_0/lib/clang/14.0.6/include/ -L/Users/ggutierr/miniconda3/pkgs/llvm-openmp-14.0.6-hc6e5704_0/lib/  main.c help.c rg.c -Xpreprocessor -fopenmp -lomp -lm -Wl,-rpath,/Users/ggutierr/miniconda3/pkgs/llvm-openmp-14.0.6-hc6e5704_0/lib/`
However, the ./a.out step isn't working and it is probably for the same reason but I haven't figured out a work around. I might just have to install Homebrew and then install libomp with it so that the libraries are where they should be. Or I can write a Make file to point things to the right directories and libraries.

To read the modularity data back into Python, we can use preprocess.py from Alex's repo. Or Rhessa made a script called process.py that might do this.
Rhessa first combined modularity data into an xls spreadsheet and then runs process.py in terminal on that and other modularity files.

It might not be necessary to run this script if we just import the data from a csv into a dataframe, but I'm not sure.